substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,814 @@
+	fn do_migrate_to_agent(who: &T::AccountId, reward_account: &T::AccountId) -> DispatchResult {
+		Self::do_register_agent(who, reward_account);
+
+		// We create a proxy delegator that will keep all the delegation funds until funds are
+		// transferred to actual delegator.
+		let proxy_delegator = Self::sub_account(AccountType::ProxyDelegator, who.clone());
+
+		// Keep proxy delegator alive until all funds are migrated.
+		frame_system::Pallet::<T>::inc_providers(&proxy_delegator);
+
+		// Get current stake
+		let stake = T::CoreStaking::stake(who)?;
+
+		// release funds from core staking.
+		T::CoreStaking::migrate_to_virtual_staker(who);
+
+		// transfer just released staked amount plus any free amount.
+		let amount_to_transfer =
+			T::Currency::reducible_balance(who, Preservation::Expendable, Fortitude::Polite);
+
+		// This should never fail but if it does, it indicates bad state and we abort.
+		T::Currency::transfer(who, &proxy_delegator, amount_to_transfer, Preservation::Expendable)?;
+
+		T::CoreStaking::update_payee(who, reward_account)?;
+		// delegate all transferred funds back to agent.
+		Self::do_delegate(&proxy_delegator, who, amount_to_transfer)?;
+
+		// if the transferred/delegated amount was greater than the stake, mark the extra as
+		// unclaimed withdrawal.
+		let unclaimed_withdraws = amount_to_transfer
+			.checked_sub(&stake.total)
+			.defensive_ok_or(ArithmeticError::Underflow)?;
+
+		if !unclaimed_withdraws.is_zero() {
+			let mut ledger = AgentLedger::<T>::get(who).ok_or(Error::<T>::NotAgent)?;
+			ledger.unclaimed_withdrawals = ledger
+				.unclaimed_withdrawals
+				.checked_add(&unclaimed_withdraws)
+				.defensive_ok_or(ArithmeticError::Overflow)?;
+			ledger.update(who);
+		}
+
+		Ok(())
+	}
+
+	/// Bond `amount` to `agent_acc` in [`Config::CoreStaking`].
+	fn do_bond(agent_acc: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		let agent = Agent::<T>::get(agent_acc)?;
+
+		let available_to_bond = agent.available_to_bond();
+		defensive_assert!(amount == available_to_bond, "not expected value to bond");
+
+		if agent.is_bonded() {
+			T::CoreStaking::bond_extra(&agent.key, amount)
+		} else {
+			T::CoreStaking::virtual_bond(&agent.key, amount, agent.reward_account())
+		}
+	}
+
+	/// Delegate `amount` from `delegator` to `agent`.
+	fn do_delegate(
+		delegator: &T::AccountId,
+		agent: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> DispatchResult {
+		let mut ledger = AgentLedger::<T>::get(agent).ok_or(Error::<T>::NotAgent)?;
+		// try to hold the funds.
+		T::Currency::hold(&HoldReason::StakingDelegation.into(), delegator, amount)?;
+
+		let new_delegation_amount =
+			if let Some(existing_delegation) = Delegation::<T>::get(delegator) {
+				ensure!(&existing_delegation.agent == agent, Error::<T>::InvalidDelegation);
+				existing_delegation
+					.amount
+					.checked_add(&amount)
+					.ok_or(ArithmeticError::Overflow)?
+			} else {
+				amount
+			};
```

Comment:
```suggestion
		defensive_assert_eq!(released, amount, "hold should have been released fully");
```

Reply:
probably commented at the wrong line? Will use `defensive_assert_eq` in places where it is checked.

Reply:
Did you comment on the wrong line?